### PR TITLE
Implement all the members in System.ComponentModel that are required for netstandard2.0

### PR DESCRIPTION
**Pull request by AlexGhiondea**

### CODE REVIEW

**Comment by AlexGhiondea** [comment]
/cc @stephentoub @twsouthwick -- this will fix the 2 issues in #12372 but I am not going to be able to fix those until this is merged and another package build. 
Will keep #12372 active until that is the case.

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/DescriptionAttribute.cs [hunk 1]
```diff
@@ -59,5 +59,10 @@ public override int GetHashCode()
         {
             return Description.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.Equals(DescriptionAttribute.Default);
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/DescriptionAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/DesignOnlyAttribute.cs [hunk 2]
```diff
@@ -70,5 +70,10 @@ public override int GetHashCode()
         {
             return IsDesignOnly.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.IsDesignOnly == DesignOnlyAttribute.Default.IsDesignOnly;
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/DesignOnlyAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/DisplayNameAttribute.cs [hunk 3]
```diff
@@ -59,5 +59,10 @@ public override int GetHashCode()
         {
             return DisplayName.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.Equals(DisplayNameAttribute.Default);
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/DisplayNameAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/LocalizableAttribute.cs [hunk 4]
```diff
@@ -68,5 +68,10 @@ public override int GetHashCode()
         {
             return base.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.IsLocalizable == LocalizableAttribute.Default.IsLocalizable;
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/LocalizableAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/MergablePropertyAttribute.cs [hunk 5]
```diff
@@ -80,5 +80,10 @@ public override int GetHashCode()
         {
             return base.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.Equals(MergablePropertyAttribute.Default);
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/MergablePropertyAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/NotifyParentPropertyAttribute.cs [hunk 6]
```diff
@@ -75,5 +75,10 @@ public override int GetHashCode()
         {
             return base.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.Equals(NotifyParentPropertyAttribute.Default);
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/NotifyParentPropertyAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/ParenthesizePropertyNameAttribute.cs [hunk 7]
```diff
@@ -71,5 +71,10 @@ public override int GetHashCode()
         {
             return base.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.Equals(ParenthesizePropertyNameAttribute.Default);
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/ParenthesizePropertyNameAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/ReadOnlyAttribute.cs [hunk 8]
```diff
@@ -77,5 +77,10 @@ public override int GetHashCode()
         {
             return base.GetHashCode();
         }
+
+        public override bool IsDefaultAttribute()
+        {
+            return this.IsReadOnly == ReadOnlyAttribute.Default.IsReadOnly;
```
Comment (justinvp) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/ReadOnlyAttribute.cs]: Nit: `this.` is unnecessary and can be removed

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 9]
```diff
@@ -12,7 +12,7 @@ namespace System.ComponentModel
     /// <summary>
     ///     Represents a collection of attributes.
     /// </summary>
-    public class AttributeCollection : ICollection
+    public class AttributeCollection : ICollection, IEnumerable
```
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: Nit: Specifying `, IEnumerable` isn't necessary

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 10]
```diff
@@ -536,5 +536,21 @@ protected static ISite GetSite(object component)
 
             return ((IComponent)component).Site;
         }
+
+        [Obsolete("This method has been deprecated. Use GetInvocationTarget instead.  http://go.microsoft.com/fwlink/?linkid=14202")]
+        protected static object GetInvokee(Type componentClass, object component) {
+
+            if (componentClass == null)
+            {
+                throw new ArgumentNullException("componentClass");
```
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Nit: `nameof(componentClass)`

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 11]
```diff
@@ -536,5 +536,21 @@ protected static ISite GetSite(object component)
...
+            }
+
+            if (component == null)
+            {
+                throw new ArgumentNullException("component");
```
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Nit: `nameof(component)`

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs [hunk 12]
```diff
@@ -378,6 +378,19 @@ object ICollection.SyncRoot
             }
         }
 
+        int ICollection.Count
```
Comment (AlexGhiondea) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: I had the same reaction about these explicit interfaces and chatted with @weshaggard about it. We decided to move them over since they are part of netstandard 2.0.

They don't add any value over their implicitly implemented counterparts but they help us reach parity with the current definition of netstandard2.0

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 13]
```diff
@@ -2637,6 +2724,180 @@ public static void SortDescriptorArray(IList infos)
...
+            ///     Implements GetTypeDescriptor.  This creates a custom type
+            ///     descriptor that walks the linked list for each of its calls.
+            //// </summary>
+            
+            [SuppressMessage("Microsoft.Globalization", "CA1303:DoNotPassLiteralsAsLocalizedParameters")]
+            public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object instance)
+            {
+                if (objectType == null)
+                {
+                    throw new ArgumentNullException("objectType");
```
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Nit: `nameof(objectType)`

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 14]
```diff
@@ -2637,6 +2724,180 @@ public static void SortDescriptorArray(IList infos)
...
+                }
+
+                if (instance == null)
+                {
+                    return null;
+                }
+
+                if (!objectType.IsInstanceOfType(instance))
+                {
+                    throw new ArgumentException("instance");
```
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Nit: `nameof(instance)`...

Also note: The first parameter to this `ArgumentException` ctor is actually for the message. The second param of `ArgumentException(string, string)` is for the `paramName`. Is there a message that could be passed as well as the `paramName`?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 15]
```diff
@@ -2637,6 +2724,180 @@ public static void SortDescriptorArray(IList infos)
...
+                internal ComNativeTypeDescriptor(IComNativeDescriptorHandler handler, object instance)
+                {
+                    _handler = handler;
+                    _instance = instance;
+                }
+#pragma warning restore 618
+
+                /// <summary>
+                ///     ICustomTypeDescriptor implementation.
+                //// </summary>
```
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Nit: All these `<summary>ICustomTypeDescriptor implementation.</summary>` don't really provide much value. Could they all be deleted?

File: src/System.ComponentModel.Primitives/src/System/ComponentModel/ReadOnlyAttribute.cs (hunk 8 above)
Comment (AlexGhiondea) [inline src/System.ComponentModel.Primitives/src/System/ComponentModel/ReadOnlyAttribute.cs]: I removed this. from all of System.ComponentModel.Primitives and TypeConverter.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 9 above)
Comment (AlexGhiondea) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: I thought about that... and I kept it there since we are also explicitly implementing those members...

I am fine to remove it if you think it does not add any value to have it there.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 12 above)
Comment (justinvp) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: What value is there in having them in netstandard 2.0? Should they be removed from netstandard 2.0?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs (hunk 15 above)
Comment (AlexGhiondea) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Gone!

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs (hunk 14 above)
Comment (AlexGhiondea) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: I guess I can do something like:
throw new ArgumentException(SR.Format(SR.ConvertToException, nameof(objectType), instance.GetType()) , nameof(instance));

Which says: 
{0} cannot convert to{1}

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 12 above)
Comment (AlexGhiondea) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: One of the [requirements](https://github.com/dotnet/standard/tree/master/docs/netstandard-20#requirements) for netstandard2.0 is that we are going to bring back types wholesale and not break them up. If we believe the type is meaningful for netstandard2.0 we have to port it over with all its members.

In this case, the types are in netstandard2.0 but they were missing some APIs (i.e. explicit interface implementations).

**Comment by AlexGhiondea** [comment]
@dotnet-bot please test Innerloop CentOS7.1 Release Build and Test (machine issue)

**Comment by AlexGhiondea** [comment]
@dotnet-bot please test Innerloop CentOS7.1 Release Build and Test (machine issue)

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 12 above)
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: Having spent time in codebase of this assembly, this assembly has many of these kinds of things; ie APIs that don't really make sense or not following current conventions. Unfortunately, as @AlexGhiondea said, as part of .NET Standard, these will be brought over wholesale, for better or worse.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindableAttribute.cs [hunk 16]
```diff
@@ -145,7 +145,7 @@ public override int GetHashCode()
         /// <internalonly/>
         public override bool IsDefaultAttribute()
         {
-            return (this.Equals(Default) || _isDefault);
+            return (Equals(Default) || _isDefault);
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BindableAttribute.cs]: These parentheses are redundant (there are other instances, but I'll just leave the one comment)

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/InvalidAsynchronousStateException.cs [hunk 17]
```diff
@@ -33,5 +35,8 @@ public InvalidAsynchronousStateException(string message, Exception innerExceptio
             : base(message, innerException)
         {
         }
+
+        protected InvalidAsynchronousStateException(SerializationInfo info, StreamingContext context) : base(info, context) {
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/InvalidAsynchronousStateException.cs]: nit: formatting of brace

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs [hunk 18]
```diff
@@ -536,5 +536,21 @@ protected static ISite GetSite(object component)
 
             return ((IComponent)component).Site;
         }
+
+        [Obsolete("This method has been deprecated. Use GetInvocationTarget instead.  http://go.microsoft.com/fwlink/?linkid=14202")]
+        protected static object GetInvokee(Type componentClass, object component) {
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Obsolete methods are being included in .NET Standard 2.0?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectPropertyDescriptor.cs [hunk 19]
```diff
@@ -1256,7 +1256,7 @@ public override void SetValue(object component, object value)
         ///     other words, it indicates whether the state of the property is distinct
         ///     from when the component is first instantiated. If there is a default
         ///     value specified in this ReflectPropertyDescriptor, it will be compared against the
-        ///     property's current value to determine this.  If there is't, the
+        ///     property's current value to determine   If there is't, the
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectPropertyDescriptor.cs]: nit: Removed the '.' from the sentence

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs [hunk 20]
```diff
@@ -5,6 +5,7 @@
 using System.Collections;
 using System.Diagnostics.CodeAnalysis;
 using System.Globalization;
+using System.ComponentModel.Design;
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs]: ? Why was this needed with no other changes?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 21]
```diff
@@ -344,7 +344,7 @@ private static void CheckDefaultProvider(Type type)
                     return;
                 }
 
-                // Immediately clear this.  If we find a default provider
+                // Immediately clear   If we find a default provider
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Don't forget the '.' :)

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 22]
```diff
@@ -2637,6 +2724,144 @@ public static void SortDescriptorArray(IList infos)
         }
 
         /// <summary>
+        ///     This class is a type description provider that works with the IComNativeDescriptorHandler
+        ///     interface.
+        //// </summary>
+        private sealed class ComNativeDescriptionProvider : TypeDescriptionProvider
+        {
+#pragma warning disable 618
+            private IComNativeDescriptorHandler _handler;
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: An auto-property would be cleaner here (and possibly a readonly auto-property)

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 23]
```diff
@@ -2637,6 +2724,144 @@ public static void SortDescriptorArray(IList infos)
...
+            }
+
+            /// <summary>
+            ///     This type descriptor sits on top of a native
+            ///     descriptor handler.
+            //// </summary>
+            private sealed class ComNativeTypeDescriptor : ICustomTypeDescriptor
+            {
+#pragma warning disable 618
+                private IComNativeDescriptorHandler _handler;
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: These can be readonly

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 24]
```diff
@@ -2637,6 +2724,144 @@ public static void SortDescriptorArray(IList infos)
...
+                    _instance = instance;
+                }
+#pragma warning restore 618
+
+                AttributeCollection ICustomTypeDescriptor.GetAttributes()
+                {
+                    return _handler.GetAttributes(_instance);
+                }
+
+                string ICustomTypeDescriptor.GetClassName()
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Do these need to be explicitly implemented?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs (hunk 12 above)
Comment (weshaggard) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs]: I didn't double check but if these are truly calling all the implicit members then I agree they are redundant and aren't actually needed for compat. When I generated what is in netstandard I took exactly what was in desktop and they do appear to implement some of the members both implicitly and explicitly, my best guess is that we did it one way in one version and then added the other version in a later version (crossing my fingers that there is a real reason :)). 

With that said I'm not going to complain either way as long as the API surface is compatible with .NET Standard.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs (hunk 18 above)
Comment (weshaggard) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/MemberDescriptor.cs]: Yes. See https://github.com/dotnet/standard/issues/6

**Comment by AlexGhiondea** [comment]
@dotnet-bot please test Innerloop Ubuntu14.04 Debug Build and Test (machine issue)
